internal/archive/credentials.go

```diff
@@ -0,0 +1,304 @@
+			errs = append(errs, fmt.Errorf("cannot parse credentials file: %w", err))
+		} else if !creds.Empty() {
+			break
+		}
+	}
+
+	err = errors.Join(errs...)
+	return
+}
+
+type netrcParser struct {
+	query   *credentialsQuery
+	scanner *bufio.Scanner
+	creds   *credentials
+}
+
+// findCredsInFile searches for credentials in netrc file matching query
+// and fills creds with matched credentials if there's a match. The first match
+// ends the search.
+//
+// The format of the netrc file is described in [1]. The parser is adapted from
+// the Apt parser (see [2]). When the parser is looking for a matching machine
+// declaration it disregards the current context and only considers the input
+// token. For example when given the following netrc file
+//
+//	machine http://acme.com/foo login u1 password machine
+//	machine http://acme.com/bar login u2 password p2
+//
+// and http://acme.com/bar input URL, the second line won't match, because the
+// second "machine" will be treated as start of machine declaration. This also
+// means unknown tokens are ignored, so comments are not treated specially.
+//
+// When a matching machine declaration is found the search stops on next
+// machine token or on end of file. This means that arbitrary number of login
+// and password declarations (or in fact, any tokens) can follow a machine
+// declaration. The last username and password declaration overrides the
+// previous ones. For example when given the following netrc file
+//
+//	machine http://acme.com login a foo login b password c bar login d password e
+//
+// and the input URL is http://acme.com, the matched username and password will
+// be "d" and "e" respectively. Tokens foo and bar will be ignored.
+//
+// This parser diverges from the Apt parser in the following ways:
+//  1. The port specification in machine declaration is optional whether or
+//     not a path is specified. While the Apt documentation[1] implies the
+//     same behavior, the code adheres to it only when the machine declaration
+//     does not specify a path, see line 96 in [2].
+//  2. When the input URL has unencrypted scheme and the machine declaration
+//     does not specify a scheme, it is skipped silently. The Apt parser warns
+//     the user about it, see line 113 in [2].
+//
+// References:
+//
+//	[1] https://manpages.debian.org/testing/apt/apt_auth.conf.5.en.html
+//	[2] https://salsa.debian.org/apt-team/apt/-/blob/d9039b24/apt-pkg/contrib/netrc.cc
+//	[3] https://salsa.debian.org/apt-team/apt/-/blob/4e04cbaf/methods/aptmethod.h#L560
+//	[4] https://www.gnu.org/software/inetutils/manual/html_node/The-_002enetrc-file.html
+//	[5] https://daniel.haxx.se/blog/2022/05/31/netrc-pains/
+func findCredsInFile(query *credentialsQuery, netrc io.Reader, creds *credentials) error {
+	s := bufio.NewScanner(netrc)
+	s.Split(bufio.ScanWords)
+	p := netrcParser{
+		query:   query,
+		scanner: s,
+		creds:   creds,
+	}
+	var err error
+	for state := netrcInvalid; state != nil; {
```

Comment:
Invalid is a weird first state for a state machine. netrcStart?

Reply:
I've explain why it is called _invalid_ instead of _start_ to Rafid in previous draft: https://github.com/canonical/chisel/pull/53#discussion_r1194399497 Quoting here:

> Not sure about this. The parser is in `netrcInvalid` every time it finds something that is not a machine it wants. So for
> ```
> machine foo login l password p #comment
> machine bar login L password P
> ```
> and query for https://foo it'd go
> 
> - invalid
> - machine
> - invalid
> - invalid
> - invalid
> - invalid
> - invalid
> - invalid
> - machine
> - goodmachine
> - login
> - goodmachine
> - password
> - goodmachine
> - nil
> 
> I may have missed or added some invalid states but my point is that it's not a special start state that's not ever entered again so I feel calling it start may be a bit misleading. Netrc format is weird in that you can paste lorem ipsum in it and it'll still find a valid machine after it. :-)

Does that make sense?

Reply:
IMO this is a matter of code comprehension. 

The initial state **should not** be a separate state, but more like an indication of which state the machine should start in by default, while also being a valid state that the machine can take during the parsing. 

I agree that "invalid" **should not** be used as the initial state, as it may introduce confusion or ambiguity in understanding the system's behaviour.

`netrcStart` feels a bit too specific, as I believe the machine would never end up in that state once the parsing is started.

What about something like "unknown". It is still a bit ambiguous, but at least it has a useful meaning and it is a state the system can take during the parsing, i.e. you should set the state to invalid whenever you are sure of it, but what if you find something you can't qualify? It could be a valid state or an invalid state...it just so happens that your code's logic can infer that, and thus the state could be "unknown".

I understand that in terms of functionality, both "invalid" and "unknown" will have similar outcomes since what you're looking for is a "goodmachine", but at least they provide a more meaningful overview of the overall system while also giving you a proper initial state.

wdyt?

Reply:
It's called `NO` in Apt: https://salsa.debian.org/apt-team/apt/-/blob/d9039b24/apt-pkg/contrib/netrc.cc#L46 I don't mind changing the name to whatever you desire (I'm fine with all of `NO`, `INVALID`, `UNKNOWN` and `START`), but introducing a separate _start_ state would be just copy & paste of the _invalid_ state.

Reply:
Added explicit start state.

Reply:
Removed explicit start state. Renamed `netrcInvalid` to `netrcStart`.